crypto/zuc/zuc.c

```diff
@@ -0,0 +1,302 @@
+    0x8E, 0x83, 0x77, 0x6B, 0x25, 0x05, 0x3F, 0x0C, 0x30, 0xEA, 0x70, 0xB7, 0xA1, 0xE8, 0xA9, 0x65,
+    0x8D, 0x27, 0x1A, 0xDB, 0x81, 0xB3, 0xA0, 0xF4, 0x45, 0x7A, 0x19, 0xDF, 0xEE, 0x78, 0x34, 0x60,
+};
+
+static unsigned char S1[256] = {
+    0x55, 0xC2, 0x63, 0x71, 0x3B, 0xC8, 0x47, 0x86, 0x9F, 0x3C, 0xDA, 0x5B, 0x29, 0xAA, 0xFD, 0x77,
+    0x8C, 0xC5, 0x94, 0x0C, 0xA6, 0x1A, 0x13, 0x00, 0xE3, 0xA8, 0x16, 0x72, 0x40, 0xF9, 0xF8, 0x42,
+    0x44, 0x26, 0x68, 0x96, 0x81, 0xD9, 0x45, 0x3E, 0x10, 0x76, 0xC6, 0xA7, 0x8B, 0x39, 0x43, 0xE1,
+    0x3A, 0xB5, 0x56, 0x2A, 0xC0, 0x6D, 0xB3, 0x05, 0x22, 0x66, 0xBF, 0xDC, 0x0B, 0xFA, 0x62, 0x48,
+    0xDD, 0x20, 0x11, 0x06, 0x36, 0xC9, 0xC1, 0xCF, 0xF6, 0x27, 0x52, 0xBB, 0x69, 0xF5, 0xD4, 0x87,
+    0x7F, 0x84, 0x4C, 0xD2, 0x9C, 0x57, 0xA4, 0xBC, 0x4F, 0x9A, 0xDF, 0xFE, 0xD6, 0x8D, 0x7A, 0xEB,
+    0x2B, 0x53, 0xD8, 0x5C, 0xA1, 0x14, 0x17, 0xFB, 0x23, 0xD5, 0x7D, 0x30, 0x67, 0x73, 0x08, 0x09,
+    0xEE, 0xB7, 0x70, 0x3F, 0x61, 0xB2, 0x19, 0x8E, 0x4E, 0xE5, 0x4B, 0x93, 0x8F, 0x5D, 0xDB, 0xA9,
+    0xAD, 0xF1, 0xAE, 0x2E, 0xCB, 0x0D, 0xFC, 0xF4, 0x2D, 0x46, 0x6E, 0x1D, 0x97, 0xE8, 0xD1, 0xE9,
+    0x4D, 0x37, 0xA5, 0x75, 0x5E, 0x83, 0x9E, 0xAB, 0x82, 0x9D, 0xB9, 0x1C, 0xE0, 0xCD, 0x49, 0x89,
+    0x01, 0xB6, 0xBD, 0x58, 0x24, 0xA2, 0x5F, 0x38, 0x78, 0x99, 0x15, 0x90, 0x50, 0xB8, 0x95, 0xE4,
+    0xD0, 0x91, 0xC7, 0xCE, 0xED, 0x0F, 0xB4, 0x6F, 0xA0, 0xCC, 0xF0, 0x02, 0x4A, 0x79, 0xC3, 0xDE,
+    0xA3, 0xEF, 0xEA, 0x51, 0xE6, 0x6B, 0x18, 0xEC, 0x1B, 0x2C, 0x80, 0xF7, 0x74, 0xE7, 0xFF, 0x21,
+    0x5A, 0x6A, 0x54, 0x1E, 0x41, 0x31, 0x92, 0x35, 0xC4, 0x33, 0x07, 0x0A, 0xBA, 0x7E, 0x0E, 0x34,
+    0x88, 0xB1, 0x98, 0x7C, 0xF3, 0x3D, 0x60, 0x6C, 0x7B, 0xCA, 0xD3, 0x1F, 0x32, 0x65, 0x04, 0x28,
+    0x64, 0xBE, 0x85, 0x9B, 0x2F, 0x59, 0x8A, 0xD7, 0xB0, 0x25, 0xAC, 0xAF, 0x12, 0x03, 0xE2, 0xF2,
+};
+
+/* D */
+static uint32_t D[16] = {
+    0x44D7, 0x26BC, 0x626B, 0x135E, 0x5789, 0x35E2, 0x7135, 0x09AF,
+    0x4D78, 0x2F13, 0x6BC4, 0x1AF1, 0x5E26, 0x3C4D, 0x789A, 0x47AC,
+};
+
+/*
+ * This is a method to calculate a + b mod (2 ^ 32 -1),
+ * described in ZUC specification.
+ */
+static ossl_inline uint32_t modular_add(uint32_t a, uint32_t b)
+{
+    uint32_t c = a + b;
+
+    return (c & 0x7FFFFFFF) + (c >> 31);
+}
+
+static ossl_inline uint32_t mulp2(uint32_t a, uint32_t b)
+{
+    return ((a << b) | (a >> (31 - b))) & 0x7FFFFFFF;
+}
+
+/* LFSR with initialization mode */
+static void zuc_lfsr_init_mode(ZUC_KEY *zk, uint32_t u)
+{
+    uint32_t tmp, v, s16;
+
+    v = mulp2(zk->s15, 15);
+    tmp = mulp2(zk->s13, 17);
+    v = modular_add(tmp, v);
+    tmp = mulp2(zk->s10, 21);
+    v = modular_add(tmp, v);
+    tmp = mulp2(zk->s4, 20);
+    v = modular_add(tmp, v);
+    tmp = mulp2(zk->s0, 8);
+    v = modular_add(tmp, v);
+    v = modular_add(zk->s0, v);
+
+    /* s16... */
+    s16 = modular_add(v, u);
+
+    /* if s16 == 0, set it to 2 ^ 31 - 1 */
```

Comment:
Ohh, I didn't consider this, I just followed the steps in the spec...But why the specs described this step?

Reply:
> But why the specs described this step?

Because needing or not needing this branch statement depends on your implementation of "mod p" according to the ZUC specification. Think about what your `modular_add` is actually doing.

Reply:
Ahh, I get it. Because the implementation of `modular_add` will directly set the addition result to modulus p (2^31 -1) if the 'addition result mod p' equals to zero.

Reply:
That's pretty much it. IMO worth documenting `modular_add` with something like "Inputs `a` and `b` are in the interval MIN to MAX because FOO. The return value is in the range MINC to MAXC because BAR.